File: examples/trigger_0x501.cpp
#include <chrono>
#include <cstdio>
#include <thread>

#include "canfd/canfd.hpp"

namespace {

constexpr uint32_t kBitrate = 1'000'000;
constexpr double kSamplePoint = 0.80;
constexpr uint32_t kDataBitrate = 5'000'000;
constexpr double kDataSamplePoint = 0.75;

constexpr uint32_t kTriggerId = 0x501;
constexpr uint32_t kResponseId = 0x481;

}  // namespace

int main() {
  canfd::CanFdBus bus;
  try {
    bus.open();

    canfd::BusConfig config;
    config.bitrate = kBitrate;
    config.sample_point = kSamplePoint;
    config.data_bitrate = kDataBitrate;
    config.data_sample_point = kDataSamplePoint;
    config.fd = true;
    bus.configure(config);
    std::printf("bus up: 1M/80%% + 5M/75%% CAN FD\n");

    canfd::CanFrame trigger;
    trigger.id = kTriggerId;
    trigger.fd = true;
    trigger.brs = true;
    trigger.size = 4;
    trigger.data[0] = 0x00;
    trigger.data[1] = 0x02;
    trigger.data[2] = 0x50;
    trigger.data[3] = 0x01;
    bus.send(trigger);

    auto last_repeat = std::chrono::steady_clock::now();
    canfd::CanFrame frame;
    while (true) {
      const auto now = std::chrono::steady_clock::now();
      if (now - last_repeat >= std::chrono::seconds(1)) {
        bus.send(trigger);
        last_repeat = now;
      }

      if (!bus.receive(frame, std::chrono::milliseconds(200))) {
        continue;
      }
      if (frame.echo || frame.id != kResponseId) {
        continue;
      }
      std::printf("%s\n", frame.toString().c_str());
    }
  } catch (const std::exception& exc) {
    std::fprintf(stderr, "error: %s\n", exc.what());
    bus.close();
    return 1;
  }
  return 0;
}

File: python/examples/trigger_0x501.py
"""Minimal end-to-end example: trigger a device on 0x501 and print 0x481.

Install the package first (``pip install ./python``) then run:

    python examples/trigger_0x501.py
"""

import time

from gsusb_canfd import BusConfig, CanFdBus, CanFrame

BITRATE = 1_000_000
SAMPLE_POINT = 0.80
DATA_BITRATE = 5_000_000
DATA_SAMPLE_POINT = 0.75

TRIGGER_ID = 0x501
TRIGGER_DATA = bytes([0x00, 0x02, 0x50, 0x01])
RESPONSE_ID = 0x481


def main() -> None:
    bus = CanFdBus()
    with bus:
        bus.configure(
            BusConfig(
                bitrate=BITRATE,
                sample_point=SAMPLE_POINT,
                data_bitrate=DATA_BITRATE,
                data_sample_point=DATA_SAMPLE_POINT,
                fd=True,
            )
        )
        print("bus up: 1M/80% + 5M/75% CAN FD")

        trigger = CanFrame(id=TRIGGER_ID, data=TRIGGER_DATA, fd=True, brs=True)
        bus.send(trigger)
        print(f"trigger sent: {TRIGGER_ID:03X} {' '.join(f'{b:02X}' for b in TRIGGER_DATA)}")

        last_repeat = time.time()
        last_data = None
        frames = 0
        print_at = time.time() + 2.0
        while True:
            if time.time() - last_repeat >= 1.0:
                bus.send(trigger)
                last_repeat = time.time()

            frame = bus.receive(timeout=0.2)
            if frame is None or frame.echo or frame.id != RESPONSE_ID:
                continue

            frames += 1
            if frame.data != last_data:
                last_data = frame.data
                print(frame)

            if time.time() >= print_at:
                print(f"--- {frames} frames on 0x{RESPONSE_ID:03X} ---")
                frames = 0
                print_at = time.time() + 2.0


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nbye")
